app/controllers/omniauth_sessions_controller.rb

```diff
@@ -0,0 +1,37 @@
+class OmniauthSessionsController < ApplicationController
+  skip_before_action :authenticate_request
+
+  def create
+    @user = ::UserFromOmniauth.new(request.env['omniauth.auth']).call
+
+    return unauthorized(@user.errors.full_messages.join("\n")) unless @user.persisted?
+    ::UpdateOAuthAccessDataForm.new(
+      session_params(auth_hash: request.env['omniauth.auth'])
+    ).save
+
+    authenticate
+  end
+
+  private
+
+  def authenticate
+    command = ::AuthenticateUser.call(@user, true)
+
+    if command.success?
+      render json: command.result
+    else
+      unauthorized(errors: command.errors)
+    end
+  end
```

Comment:
Can we return a useful error instead of the two json responses here? This is shown as raw data in the user's browser after they authenticate. Maybe something like `Connected (email) account` and `Error authenticating`.

Reply:
Something like :
```
if command success?
 render json: "Connected #{command[:user][:email] account"
else
 unauthorized(errors: "Error authenticating")
end
```

Reply:
Instead of json, I think we should return html or text.

```ruby
if command.success?
  render html: "Connected #{command.result[:user]['email']} account. You may now close this browser tab."
else
  render html: "Unable to connect account."
end
```

You will also need a `skip_before_action :force_json` at the top of the file to be able to use html.

Reply:
Feel free to use some sort of html here if you have a more elegant idea!